openedx/core/djangoapps/agreements/api.py

```diff
@@ -68,3 +73,165 @@ def get_integrity_signatures_for_course(course_id):
+    """
+    Get a course's LTI tools that share PII.
+
+    Arguments:
+        * course_id (str)
+
+    Returns:
+        * A List of LTI tools sharing PII.
+    """
+
+    course_key = CourseKey.from_string(course_id)
+    course_ltipiitools = LTIPIITool.objects.get(course_key=course_key).lti_tools
+
+    return LTIToolsReceivingPIIData(
+        lii_tools_receiving_pii=course_ltipiitools,
+    )
+
+
+def user_lti_pii_signature_needed(username, course_id):
+    """
+    Determines if a user needs to acknowledge the LTI PII Agreement
+
+    Arguments:
+        * username (str)
+
+    Returns:
+        * True if the user needs to sign a new acknowledgement.
+        * False if the acknowledgements are up to date.
+    """
+    if _course_has_lti_pii_tools(course_id):
+        if _user_lti_pii_signature_exists(username, course_id):
+            if _user_needs_signature_update(username, course_id):
+                # up to date
+                return False
+            else:
+                # lti pii signature needs to be updated
+                return True
+        else:
+            # write a new lti pii signature
+            return True
+    else:
+        return False
+
+
+def _course_has_lti_pii_tools(course_id):
+    """
+    Determines if a specifc course has lti tools sharing pii
+
+    Arguments:
+        * course_id (str)
+
+    Returns:
+        * True if the course does have a list.
+        * False if the course does not.
+    """
+    course_key = CourseKey.from_string(course_id)
+    course_lti_pii_tools = LTIPIITool.objects.get(course_key)
+
+    if not course_lti_pii_tools:
+        # empty queryset, meaning no tools
+        return False
+    else:
+        return True
+
+
+def _user_lti_pii_signature_exists(username, course_id):
+    """
+    Determines if a user's lti pii signature exists for a specfic course
+
+    Arguments:
+        * username (str)
+        * course_id (str)
+
+    Returns:
+        * True if user has a signature for the given course.
+        * False if the user does not have a signature for the given course.
+    """
+    user = User.objects.get(username=username)
+    course_key = CourseKey.from_string(course_id)
+
+    signature = LTIPIISignature.objects.get(user=user, course_key=course_key)
+    if not signature:
+        return False
+    else:
+        return True
+
+
+def _user_needs_signature_update(username, course_id):
+    """
+        Determines if a user's existing lti pii signature is out-of-date for a given course.
+
+        Arguments:
+            * username (str)
+            * course_id (str)
+
+        Returns:
+            * True if user has a signature for the given course.
+            * False if the user does not have a signature for the given course.
+        """
+
+    user = User.objects.get(username=username)
+    course_key = CourseKey.from_string(course_id)
+
+    user_lti_pii_signature_hash = LTIPIISignature.objects.get(course_key=course_key, user=user).lti_tools_hash
```

Comment:
Would the try-except clause include the ```user.objects.get()``` line as well? I am thinking that if a user is in the course they'd have to exist, but I'm not sure if it's safe to assume.

Reply:
That's a good question. You have a good point that we'd really only expect for these APIs to be called in the context of someone being logged in and navigating a course, in which case they would definitely exist in the database and this call would always succeed.

But, to be safe, I think it would be best to wrap this call in a try-except block as well on the off chance that these APIs are used inappropriately without a valid username. Stuff happens!